staging/src/k8s.io/apiserver/pkg/apis/apiserverinternal/v1alpha1/types.go

```diff
@@ -0,0 +1,127 @@
+
+	// Spec is an empty spec. It is here to comply with Kubernetes API style.
+	Spec StorageVersionSpec `json:"spec" protobuf:"bytes,2,opt,name=spec"`
+
+	// API server instances report the version they can decode and the version they
+	// encode objects to when persisting objects in the backend.
+	Status StorageVersionStatus `json:"status" protobuf:"bytes,3,opt,name=status"`
+}
+
+// StorageVersionSpec is an empty spec.
+type StorageVersionSpec struct{}
+
+// API server instances report the versions they can decode and the version they
+// encode objects to when persisting objects in the backend.
+type StorageVersionStatus struct {
+	// The reported versions per API server instance.
+	// +optional
+	// +listType=map
+	// +listMapKey=apiserverID
+	StorageVersions []ServerStorageVersion `json:"storageVersions,omitempty" protobuf:"bytes,1,opt,name=storageVersions"`
+	// If all API server instances agree on the same encoding storage version,
+	// then this field is set to that version. Otherwise this field is left empty.
+	// API servers should finish updating its storageVersionStatus entry before
+	// serving write operations, so that this field will be in sync with the reality.
+	// +optional
+	CommonEncodingVersion *string `json:"commonEncodingVersion,omitempty" protobuf:"bytes,2,opt,name=commonEncodingVersion"`
+
+	// The latest available observations of the storageVersion's state.
+	// +optional
+	// +listType=map
+	// +listMapKey=type
+	Conditions []StorageVersionCondition `json:"conditions,omitempty" protobuf:"bytes,3,opt,name=conditions"`
+}
+
+// An API server instance reports the version it can decode and the version it
+// encodes objects to when persisting objects in the backend.
+type ServerStorageVersion struct {
+	// The ID of the reporting API server.
+	APIServerID string `json:"apiServerID,omitempty" protobuf:"bytes,1,opt,name=apiServerID"`
+
+	// The API server encodes the object to this version when persisting it in
+	// the backend (e.g., etcd).
+	EncodingVersion string `json:"encodingVersion,omitempty" protobuf:"bytes,2,opt,name=encodingVersion"`
+
+	// The API server can decode objects encoded in these versions.
+	// The encodingVersion must be included in the decodableVersions.
+	// +listType=set
+	DecodableVersions []string `json:"decodableVersions,omitempty" protobuf:"bytes,3,opt,name=decodableVersions"`
+}
+
+type StorageVersionConditionType string
+
+const (
+	// Indicates that encoding storage versions reported by all servers are equal.
+	AllEncondingVersionsEqual StorageVersionConditionType = "AllEncodingVersionsEqual"
+)
+
+type ConditionStatus string
+
+const (
+	ConditionTrue    ConditionStatus = "True"
+	ConditionFalse   ConditionStatus = "False"
+	ConditionUnknown ConditionStatus = "Unknown"
+)
+
+// Describes the state of the storageVersion at a certain point.
+type StorageVersionCondition struct {
+	// Type of the condition.
+	// +optional
```

Comment:
`type` should be required since it's the listMapKey (also based on [this KEP](https://github.com/deads2k/enhancements/blob/master/keps/sig-api-machinery/1623-standardize-conditions/README.md#proposal))